tools/crm_mon.c

```diff
@@ -3000,21 +3006,44 @@ reduce_stonith_history(stonith_history_t *history)
     for (hp = history; hp; ) {
         for (np = new; np; np = np->next) {
             if ((hp->state == st_done) || (hp->state == st_failed)) {
+                enum redude_state reduce = reduce_none;
+
                 /* action not in progress */
                 if (safe_str_eq(hp->target, np->target) &&
-                    safe_str_eq(hp->action, np->action) &&
-                    (hp->state == np->state)) {
```

Comment:
Didn't think it through all the way but I doubt that it will work as desired.
Assume there are failed actions with different delegates in the list and then there comes a successful action. That should replace one of the failed actions and the other one should should be purged. Don't see how that would happen with this code.
Guess when we decide for purging the history entry we are fine with the current behavior. But in case we are replacing an entry in the new table we would do that but instead of breaking the loop we should go over the rest of the new table and purge all further matches.
Readability might improve if we introduce a compare function between 2 entries we would exchange dependent on if we want the old or new behavior (maybe controlled via m=0,1,2,3,4 - or thinking over it one more time as suppression of failed actions would be interesting for both current m=1 & m=2 using an additional control).

Reply:
Certainly as you say.
Think again to be processed by Purge.

Reply:
Hi wenningerk,

Purging data from the hp list is easy, but purging from the np list (new) also complicates source code.

It's a different approach, but what about a fix that adds a flag and sets this flag with reduce_stonith_history ()?

```
include/crm/stonith_ng.h
(snip)
typedef struct stonith_history_s {    char *target;    char *action;    char *origin;
    char *delegate;
    int completed;
    int state;

    struct stonith_history_s *next;
    char *client;
    gboolean ignore_failed; /* add flags */
} stonith_history_t;
```


Best Regards,
Hideo Yamauchi.

Reply:
Other approach might be using quicksort as I'm doing in the daemon-code to sort for timestamp. Afterwards we could totally skip the overwrite path and would never have the issue with a newer success entry having to overwrite one failure while multiple might have to be purged.

Reply:
Btw. maybe you can invite clumens for the review as I know he is working on crm_mon atm. Not sure if it touches this codepart - just to prevent that we introduce something that might be hard to merge.

Reply:
Hi wenningerk,

If you include quicksort (), it will make the process easier to read.
I think it's easier now to use the flag that I wrote in the second comment, so let's get PR out again.
Let's invite Clumens to a review with that PR.

I do not know ...
By the way, do you know what "crm_mon atm" means?

Best Regards,
Hideo Yamauchi.

Reply:
a(t) t(he) m(oment) ;-)

Reply:
Okay!;-)

As I do PR again, I close this PR.

Many thanks,
Hideo Yamauchi.